### ISSUE TITLE
Popup actions partially hidden

### ISSUE THREAD

**Issue by manuel-uberti** [tarsius/moody#19]
Following Magit documentation, I added some actions to the file popup with this:
```
(use-package magit-files                ; Magit file operations
  :ensure magit
  :config
  ;; Add more operations to the file popup
  (magit-define-popup-action 'magit-file-popup
    ?R "Rename file" 'magit-file-rename)
  (magit-define-popup-action 'magit-file-popup
    ?K "Delete file" 'magit-file-delete)
  (magit-define-popup-action 'magit-file-popup
    ?U "Untrack file" 'magit-file-untrack)
  (magit-define-popup-action 'magit-file-popup
    ?C "Checkout file" 'magit-file-checkout))
```
When I use <kbd>M-x magit-file-popup</kbd> this is what I see:

![popup](https://user-images.githubusercontent.com/8199224/42148189-87199f2c-7dd1-11e8-971c-88c2c0f2f541.png)

Same actions are completely visible with `moody` disabled.

This is how I configured `moody`:
```
(use-package moody                      ; Tabs and ribbons for the mode line
  :ensure t
  :config
  (moody-replace-mode-line-buffer-identification)
  (moody-replace-vc-mode))
```

I also slightly decreased the font in the mode-line, if this is relevant:
```
(set-face-attribute 'mode-line nil
                    :family "Iosevka"
                    :height 130)
```
And these are some details about my system, in case they can help with debugging:
```
GNU Emacs 27.0.50 (build 1, x86_64-debian-linux-gnu, GTK+ Version 3.22.30)
 of 2018-07-02

Repository revision: ee3e432300054ca488896e39fca57b10d733330a
Configured using:
--host=x86_64-debian-linux-gnu --with-modules

Emacs uptime: 4 minutes, 36 seconds
Colour theme: sanityinc-tomorrow-night
Operating system: Ubuntu 18.04 LTS
Window system: x11
Desktop environment: GNOME Shell 3.28.1
```

**Comment by tarsius** [comment]
I am attempting to prevent that with adf652f35cba1bb3d0f254e1905e2deeeb0fbdba, for me it has worked so far.

**Comment by manuel-uberti** [comment]
I see. There is that line in my `moody.el`, but the display glitch remains.

**Comment by manuel-uberti** [comment]
Further investigation reveals this is not happening on my Dell XPS 13 (3200x1800), but only when I plug in the external display (1920x1080).

**Comment by tarsius** [comment]
I cannot reproduce this and don't know what's going on. If you want to get that fixed you'll have to do the research yourself. I suspect some calculation in Emacs is off somehow.

**Comment by manuel-uberti** [comment]
Thank you. I'll see what I can do, but since it seems to be happening only for me I reckon we can close the issue.

**Comment by tarsius** [comment]
Closing then :wink:

**Comment by macroscian** [comment]
Appreciate this is closed, but hope it's ok to say I've been experiencing a similar problem when running `jump-to-register` which generates a pop-up list containing four entries when moody isn't in operation, but the bottom line is obscured when moody is active.   I'm on GNU Emacs 26.2 (build 1, x86_64-pc-linux-gnu, X toolkit, Xaw scroll bars) of 2019-05-07 with  moody 20190203.1747 installed.  It happens to me when I run with -q, add a few entries to my register and then enable moody: I did oddly enough first notice this running a laptop on an external display, but it doesn't seem specific to that setup for me.  Thanks for the great packages, and apologies for pointing out a trivial niggle.
 
<img width="348" alt="moody" src="https://user-images.githubusercontent.com/1587213/62364769-4a30cc80-b51a-11e9-839a-20a80453ea9f.png">

**Comment by hpdeifel** [comment]
I've got the same issue with `org-agenda-tree-to-indirect-buffer`.

**Comment by hpdeifel** [comment]
Same thing with `org-set-tags-command`, when using fast tag selection.

**Comment by tarsius** [comment]
```
Certain functions, including e.g. `fit-window-to-buffer', base
their size calculations on values that are incorrect if the
mode-line has a height different from that of the `default' face
and certain other calculations have not yet taken place for the
window in question.

Force a redisplay before `fit-window-to-buffer' using an advice,
like we already did for `resize-temp-buffer-window'.
```

This fixes the issue for org-set-tags-command. I could not check if it does so for org-agenda-tree-to-indirect-buffer because I am not using an agenda. And for me jump-to-register did not have this issue even without this advice.

If you encounter a similar issue with some other command, then follow the code until you have the determined the function that does the resizing of the relevant window. Ideally that should be some general-purpose function that is not specific to the package or even command in question. Then try if advising that function does the trick. If so, then please report that so that I can add the advise the function in `moody.el` for others to benefit.

**Comment by hpdeifel** [comment]
Seems to work for `org-set-tags-command` as well as `org-agenda-tree-to-indirect-buffer`. Thanks!

**Comment by seagle0128** [comment]
The height issue was resolved, but the workaround brings a new blinking issue as below. 

![8671d4453aec880db6fce8fa6e864940524e3252](https://user-images.githubusercontent.com/140797/72286755-be060200-3680-11ea-9313-914e260f01a5.gif)

**Comment by tarsius** [comment]
I've improved the workaround.

Also https://debbugs.gnu.org/cgi/bugreport.cgi?bug=38181#86 sounds like it might work.

**Comment by seagle0128** [comment]
It works. Hope the ultimate solution is committed soon.

**Comment by memeplex** [comment]
One thing that may be important: the first time the completion popup opens it's properly sized, it's after the first time that it gets clipped and then stays like that. So maybe the underlying reason is different, since I can't see how a missing call to redisplay won't affect the first popup but will break the layout of next one. Therefore I disabled the "just once" behavior (hacked-p) and it fixes the immediate problem. That said, I don't know the implications of redisplaying each time `fit-window-to-buffer` is called, but it's apparent that the hack stop working at some time.

**Comment by memeplex** [comment]
I've been driving without the "not hacked-p" seat belt but logging each advised redisplay and so far so good, with a mix of treemacs, company, company-box, doom and moody modelines, ivy, polymode, gnus, which-key,... the only things triggering a redisplay are that org-mode popup and which-key. I believe that doom-emacs hides the modeline before showing popups, maybe that's a good trick that even looks nicer than the current pile of interaction buffers. I also necrobumped #38181 upstream, given that it's not a just once per buffer issue it seems more serious than before.

**Comment by tarsius** [comment]
Thanks to @memeplex this is now fixed in Emacs 29.0.50 and we use a better kludge for older versions. See https://github.com/tarsius/moody/pull/35.
